# Week4授業前課題3 オブジェクト指向の活用

<h1>目次<span class="tocSkip"></span></h1>
<div class="toc"><ul class="toc-item"><li><span><a href="#このテキストについて" data-toc-modified-id="このテキストについて-1"><span class="toc-item-num">1&nbsp;&nbsp;</span>このテキストについて</a></span><ul class="toc-item"><li><span><a href="#テキストの目的" data-toc-modified-id="テキストの目的-1.1"><span class="toc-item-num">1.1&nbsp;&nbsp;</span>テキストの目的</a></span></li><li><span><a href="#どのように学ぶか" data-toc-modified-id="どのように学ぶか-1.2"><span class="toc-item-num">1.2&nbsp;&nbsp;</span>どのように学ぶか</a></span></li></ul></li><li><span><a href="#オブジェクト指向" data-toc-modified-id="オブジェクト指向-2"><span class="toc-item-num">2&nbsp;&nbsp;</span>オブジェクト指向</a></span></li><li><span><a href="#scikit-learnの標準化クラス" data-toc-modified-id="scikit-learnの標準化クラス-3"><span class="toc-item-num">3&nbsp;&nbsp;</span>scikit-learnの標準化クラス</a></span><ul class="toc-item"><li><span><a href="#インスタンス化" data-toc-modified-id="インスタンス化-3.1"><span class="toc-item-num">3.1&nbsp;&nbsp;</span>インスタンス化</a></span></li><li><span><a href="#【問題1】これまで利用してきたクラスの列挙" data-toc-modified-id="【問題1】これまで利用してきたクラスの列挙-3.2"><span class="toc-item-num">3.2&nbsp;&nbsp;</span>【問題1】これまで利用してきたクラスの列挙</a></span></li><li><span><a href="#メソッド" data-toc-modified-id="メソッド-3.3"><span class="toc-item-num">3.3&nbsp;&nbsp;</span>メソッド</a></span></li><li><span><a href="#インスタンス変数（アトリビュート）" data-toc-modified-id="インスタンス変数（アトリビュート）-3.4"><span class="toc-item-num">3.4&nbsp;&nbsp;</span>インスタンス変数（アトリビュート）</a></span></li><li><span><a href="#【問題2】これまで利用してきたメソッドやインスタンス変数の列挙" data-toc-modified-id="【問題2】これまで利用してきたメソッドやインスタンス変数の列挙-3.5"><span class="toc-item-num">3.5&nbsp;&nbsp;</span>【問題2】これまで利用してきたメソッドやインスタンス変数の列挙</a></span></li><li><span><a href="#インスタンス変数をメソッドが利用" data-toc-modified-id="インスタンス変数をメソッドが利用-3.6"><span class="toc-item-num">3.6&nbsp;&nbsp;</span>インスタンス変数をメソッドが利用</a></span></li><li><span><a href="#【問題3】標準化クラスをスクラッチで作成" data-toc-modified-id="【問題3】標準化クラスをスクラッチで作成-3.7"><span class="toc-item-num">3.7&nbsp;&nbsp;</span>【問題3】標準化クラスをスクラッチで作成</a></span></li><li><span><a href="#ライブラリのソースコードを確認" data-toc-modified-id="ライブラリのソースコードを確認-3.8"><span class="toc-item-num">3.8&nbsp;&nbsp;</span>ライブラリのソースコードを確認</a></span></li><li><span><a href="#特殊メソッド" data-toc-modified-id="特殊メソッド-3.9"><span class="toc-item-num">3.9&nbsp;&nbsp;</span>特殊メソッド</a></span></li><li><span><a href="#【問題4】-四則演算を行うクラスの作成" data-toc-modified-id="【問題4】-四則演算を行うクラスの作成-3.10"><span class="toc-item-num">3.10&nbsp;&nbsp;</span>【問題4】 四則演算を行うクラスの作成</a></span></li></ul></li></ul></div>

## このテキストについて

### テキストの目的
* クラスを利用したコードを読み書きできるようにする

### どのように学ぶか
* これまで使用してきたクラスを元にしてオブジェクト指向を理解していきます。

## オブジェクト指向

これまでの課題では触れてきませんでしたが、StandardScalerやLinearRegressionのようなクラスと呼ばれるものがPythonなどのプログラム言語では利用できます。  

クラスの構文は、オブジェクト指向と呼ばれる考え方を利用したプログラミングの基本的な道具になります。

この課題ではこれまでに既に登場していたクラスを例に、クラスを活用することでどのようなことができるのかを見て学んでいきます。そして課題の後半ではStandardScalerのクラスをスクラッチで自作します。

## scikit-learnの標準化クラス
scikit-learnに用意されている標準化を行うためのクラスStandardScalerを例に見ていきます。  
サンプルコードを用意しましたので、これを利用しながら理解していきます。  
  
sklearn.preprocessing.StandardScaler — scikit-learn 0.21.3 documentation  
http://scikit-learn.org/stable/modules/generated/sklearn.preprocessing.StandardScaler.html

In [68]:
# 《サンプルコード》

import numpy as np
from sklearn.preprocessing import StandardScaler
from sklearn.datasets import load_iris

data = load_iris()
X = data.data[:10]
scaler = StandardScaler()
scaler.fit(X)
print("平均 :", scaler.mean_)
print("分散 :", scaler.var_)
X_std = scaler.transform(X)

平均 : [4.86 3.31 1.45 0.22]
分散 : [0.0764 0.0849 0.0105 0.0056]


### インスタンス化
クラスを使う際はまず以下のようなコードを書きますが、これをインスタンス化と呼びます。

scaler = StandardScaler()

StandardScalerというクラスオブジェクトから、scalerと名前をつけたインスタンスオブジェクトが作られました。

《クラスの命名法》

Pythonではクラス名は頭文字が大文字、他は小文字という命名法がPEP8により定められています。  
単語間にアンダースコアは入れません。これを CapWords方式と呼びます。

はじめに — pep8-ja 1.0 ドキュメント クラスの名前  
https://pep8-ja.readthedocs.io/ja/latest/#id31

こういった形式のものはクラスだと判断することができます。

《インスタンスは複数作れる》

あるクラスオブジェクトからは複数のインスタンスオブジェクトを作成することが可能です。
1. scaler0 = StandardScaler()
2. scaler1 = StandardScaler()
3. scaler2 = StandardScaler()

### 【問題1】これまで利用してきたクラスの列挙
クラスを使う際はインスタンス化を行うことと、クラスの命名法がわかりました。  
この情報を元に、これまでの課題で利用してきたコードの中でどのようなクラスがあったかを答えてください。  
  
Pandas、matplotlib、scikit-learnからそれぞれ1つ以上見つけてください。

* Pandas・・・DataFrame()
* matplotlib・・・Circle()
* scikit-learn・・・LogisticRegression()

### メソッド
インスタンス化を行った後には、scaler.fit(X)のようなメソッドの実行ができます。  
StandardScalerのfitメソッドは後でスケーリングに使われる平均と標準偏差を計算する機能があります。

### インスタンス変数（アトリビュート）
fitメソッドにより平均と標準偏差が計算されましたが、見た目には変化があるわけではありません。  
しかし、scalerインスタンスの内部では計算結果が保存されています。  
  
こういったインスタンスの中で値を保存するものをインスタンス変数やアトリビュート（属性）と呼びます。  
ここで平均がscaler.mean_、標準偏差の2乗した値である分散がscaler.var_に保存されています。

以下のようにprint文で出力させることができます。
1. print("平均 : {}".format(scaler.mean_)) # 平均 : [4.86 3.31 1.45 0.22]  
2. print("分散 : {}".format(scaler.var_)) # 分散 : [0.0764 0.0849 0.0105 0.0056]  

《メソッドとインスタンス変数の命名法》

メソッドやインスタンス変数の命名は関数と同様に、全て小文字で行います。  
単語をつなぐときにはアンダースコアを入れます。  
  
はじめに — pep8-ja 1.0 ドキュメント メソッド名とインスタンス変数  
https://pep8-ja.readthedocs.io/ja/latest/#id37

### 【問題2】これまで利用してきたメソッドやインスタンス変数の列挙
これまでの課題で利用してきたコードの中でどのようなメソッドやインスタンス変数があったかを答えてください。
最低でもそれぞれ5つ以上答えてください。
  
＜回答＞  
メソッド
* format()
* append()
* int()
* info()
* sort()
* items()

インスタンス変数
* shape
* columns
* size
* T
* coef_
  
《ndarrayやstrもインスタンス》

ドットをつけるというと、NumPyのndarrayに対してndarray.shapeやndarray.sum()のような使い方は何度も利用してきたかと思います。これは、ndarrayもインスタンスオブジェクトであり、shapeはインスタンス変数、sumはメソッドだったということです。

Pythonのコードに登場するデータはどれもインスタンスオブジェクトであり、listやstrもメソッドを持ちます。

（例）  
5. データ構造 — Python 3.7.4 ドキュメント 5.1. リスト型についてもう少し  
https://docs.python.jp/3/tutorial/datastructures.html#more-on-lists  

* l = ['a']
* l.append('b') # listのappendメソッド
  
4. 組み込み型 — Python 3.7.4 ドキュメント 文字列メソッド  
https://docs.python.jp/3/library/stdtypes.html#string-methods

* s = 'Hello, World!'
* s.find('W') # strのfindメソッド

### インスタンス変数をメソッドが利用
最終的に以下のようにして標準化を行います。  
  
X_std = scaler.transform(X)  
  
これはfitメソッドで計算したことでインスタンス変数mean_やvar_に保存されていた値を使い、  
Xを変換したということです。
  
このようにクラスには複数のメソッドやインスタンス変数が存在し、これらを組み合わせていろいろな機能を実現します。  

### 【問題3】標準化クラスをスクラッチで作成
理解をより深めるため、StandardScalerをスクラッチで作成しましょう。  
scikit-learnは使わず、NumPyなどを活用して標準化の計算を記述します。  
具体的にはfitメソッドとtransformメソッドを作ります。  
  
今回は雛形を用意しました。クラスの作成方法は関数に近いです。  
メソッドはクラスの中にさらにインデントを一段下げて記述します。  

インスタンス変数を作成する際はself.mean_のようにselfを付けます。  
クラスの外からscaler.mean_と書いていたscalerの部分が自分自身を表すselfになっています。

In [69]:
# スクラッチする
class ScratchStandardScaler():
    '''
    標準化のためのクラス
    
    Attributes
    ----------
    mean_ : 次の形のndarray, shape(n_features,), 平均
    var_ : 次の形のndarray, shape(n_features,), 分散
    '''
    
    def fit(self,X):
        """
        標準化のために平均と標準偏差を計算する。

        Parameters
        ----------
        X : 次の形のndarray, shape (n_samples, n_features), 訓練データ
        """
        
        self.mean_ = np.mean(X,axis=0)
        
        self.var_ = np.var(X,axis=0)
        pass
        
    def transform(self,X):
        """
        fitで求めた値を使い標準化を行う。

        Parameters
        ----------
        X : 次の形のndarray, shape (n_samples, n_features)
            特徴量

        Returns
        ----------
        X_scaled : 次の形のndarray, shape (n_samples, n_features)
            標準化された特徴量
        """
        
        X_scaled = (X - self.mean_)/np.sqrt(self.var_)
        
        pass
        return X_scaled

In [78]:
# 標準化の計算過程を表示
print('ndarray : X  \n',X)

print('(X - X.mean)  \n',(X - scaler.mean_))

print('(X - X.mean)/X.std  \n',(X - scaler.mean_)/np.sqrt(scaler.var_))

ndarray : X  
 [[5.1 3.5 1.4 0.2]
 [4.9 3.  1.4 0.2]
 [4.7 3.2 1.3 0.2]
 [4.6 3.1 1.5 0.2]
 [5.  3.6 1.4 0.2]
 [5.4 3.9 1.7 0.4]
 [4.6 3.4 1.4 0.3]
 [5.  3.4 1.5 0.2]
 [4.4 2.9 1.4 0.2]
 [4.9 3.1 1.5 0.1]]
(X - X.mean)  
 [[ 0.24  0.19 -0.05 -0.02]
 [ 0.04 -0.31 -0.05 -0.02]
 [-0.16 -0.11 -0.15 -0.02]
 [-0.26 -0.21  0.05 -0.02]
 [ 0.14  0.29 -0.05 -0.02]
 [ 0.54  0.59  0.25  0.18]
 [-0.26  0.09 -0.05  0.08]
 [ 0.14  0.09  0.05 -0.02]
 [-0.46 -0.41 -0.05 -0.02]
 [ 0.04 -0.21  0.05 -0.12]]
(X - X.mean)/X.std  
 [[ 0.86828953  0.65207831 -0.48795004 -0.26726124]
 [ 0.14471492 -1.06391725 -0.48795004 -0.26726124]
 [-0.57885968 -0.37751902 -1.46385011 -0.26726124]
 [-0.94064699 -0.72071813  0.48795004 -0.26726124]
 [ 0.50650222  0.99527742 -0.48795004 -0.26726124]
 [ 1.95365143  2.02487476  2.43975018  2.40535118]
 [-0.94064699  0.3088792  -0.48795004  1.06904497]
 [ 0.50650222  0.3088792   0.48795004 -0.26726124]
 [-1.66422159 -1.40711636 -0.48795004 -0.26726124]
 [ 0.14471492 -0.72071813 

In [73]:
ScratchScaler = ScratchStandardScaler()

ScratchScaler.fit(X)
print("平均 :", ScratchScaler.mean_)
print("分散 :", ScratchScaler.var_)

Scratch_X_std = ScratchScaler.transform(X)
print(Scratch_X_std)

平均 : [4.86 3.31 1.45 0.22]
分散 : [0.0764 0.0849 0.0105 0.0056]
[[ 0.86828953  0.65207831 -0.48795004 -0.26726124]
 [ 0.14471492 -1.06391725 -0.48795004 -0.26726124]
 [-0.57885968 -0.37751902 -1.46385011 -0.26726124]
 [-0.94064699 -0.72071813  0.48795004 -0.26726124]
 [ 0.50650222  0.99527742 -0.48795004 -0.26726124]
 [ 1.95365143  2.02487476  2.43975018  2.40535118]
 [-0.94064699  0.3088792  -0.48795004  1.06904497]
 [ 0.50650222  0.3088792   0.48795004 -0.26726124]
 [-1.66422159 -1.40711636 -0.48795004 -0.26726124]
 [ 0.14471492 -0.72071813  0.48795004 -1.60356745]]


In [74]:
scaler = StandardScaler()
scaler.fit(X)
print("平均 :", scaler.mean_)
print("分散 :", scaler.var_)
X_std = scaler.transform(X)
print(X_std)

平均 : [4.86 3.31 1.45 0.22]
分散 : [0.0764 0.0849 0.0105 0.0056]
[[ 0.86828953  0.65207831 -0.48795004 -0.26726124]
 [ 0.14471492 -1.06391725 -0.48795004 -0.26726124]
 [-0.57885968 -0.37751902 -1.46385011 -0.26726124]
 [-0.94064699 -0.72071813  0.48795004 -0.26726124]
 [ 0.50650222  0.99527742 -0.48795004 -0.26726124]
 [ 1.95365143  2.02487476  2.43975018  2.40535118]
 [-0.94064699  0.3088792  -0.48795004  1.06904497]
 [ 0.50650222  0.3088792   0.48795004 -0.26726124]
 [-1.66422159 -1.40711636 -0.48795004 -0.26726124]
 [ 0.14471492 -0.72071813  0.48795004 -1.60356745]]


### ライブラリのソースコードを確認
scikit-learnの場合は公式ドキュメントの右上にソースコードへのリンクがあります。

https://github.com/scikit-learn/scikit-learn/blob/7389dba/sklearn/preprocessing/data.py#L480

どのようなコードになっていたかを確認してみましょう。（問題3に取り組んだ後に見ることを推奨します）  
スクラッチで作成したものよりも全体的にコードが長いのではないかと思います。  
inverse_transformメソッドのように作成しなかったものもありますが、それだけではありません。  
例えば以下のように、warning文が記述されているなどします。

しかし、特に今注目したいのは次の特殊メソッドについてです。

### 特殊メソッド
ソースコードの中に含まれる、まだ説明していない重要な部分が以下です。  
このような__init__というメソッドは、どのクラスにも共通して置かれるコンストラクタと呼ばれるメソッドです。  

def __init__(self, copy=True, with_mean=True, with_std=True):  
    self.with_mean = with_mean  
    self.with_std = with_std  
    self.copy = copy  

今回のスクラッチではcopy、with_mean、with_stdなどのパラメータを省略しましたが、このようにインスタンス化の際にパラメータを指定して保存しておくということはよくある使い方です。

コンストラクタの動作を確認するためのサンプルコードを用意しました。コンストラクタは、インスタンス化が行われる時に自動的に実行されるという働きがあります。こういった特殊な動作をするメソッドを、 特殊メソッド と呼びます。

《サンプルコード》

class ExampleClass():
    """
    説明用の簡単なクラス

    Parameters
    ----------
    value : float or int
        初期値

    Attributes
    ----------
    value : float or int
        計算結果
    """
    def __init__(self, value):
        self.value = value
        print("初期値{}が設定されました".format(self.value))
    def add(self, value2):
        """
        受け取った引数をself.valueに加える
        """
        self.value += value2
example = ExampleClass(5)  
print("value : {}".format(example.value))  
example.add(3)  
print("value : {}".format(example.value))  

### 【問題4】 四則演算を行うクラスの作成
上記ExampleClassは足し算のメソッドを持っていますが、これに引き算、掛け算、割り算のメソッドを加えてください。

コンストラクタに入力されたvalueが文字列や配列など数値以外だった場合にはエラーを出すようにしてください。

クラス名や説明文も適切に書き換えてください。

In [137]:
class FourArithmeticOperations():
    """
    説明用の簡単なクラス

    Parameters
    ----------
    value : float or int
        初期値

    Attributes
    ----------
    value : float or int
        計算結果
    """
    
    def __init__(self, value):
        self.value = value
        
        assert type(self.value) == (int or float),\
        'Can not calcuration, Please input int or float'
        
        print("初期値{}が設定されました".format(self.value))
                
    def add(self, value2):
        """
        受け取った引数をself.valueに加える
        """
        assert type(value2) == (int or float),\
        'Can not calcuration, Please input int or float'
        
        self.value += value2
        
    def sub(self, value2):
        """
        受け取った引数をself.valueから引く
        """
        assert type(value2) == (int or float),\
        'Can not calcuration, Please input int or float'
        
        self.value -= value2
        
    def mult(self, value2):
        """
        受け取った引数をself.valueに掛ける
        """
        assert type(value2) == (int or float),\
        'Can not calcuration, Please input int or float'
        
        self.value *= value2
        
    def divi(self, value2):
        """
        受け取った引数でself.valueを割る
        """
        assert type(value2) == (int or float),\
        'Can not calcuration, Please input int or float'
        
        self.value /= value2
        
            
initial = 1
            
cal = FourArithmeticOperations(initial)
print("value : {}".format(cal.value))
cal.add(3)
print("value : {}".format(cal.value))
cal.sub(100)
print("value : {}".format(cal.value))
cal.mult(3)
print("value : {}".format(cal.value))
cal.divi(90)
print("value : {}".format(cal.value))

初期値1が設定されました
value : 1
value : 4
value : -96
value : -288
value : -3.2
